### PR TITLE
Added login/register UI

### LINKED ISSUE #676: creating new password error handling
As a user creating a new account with email and password, I have to enter a password that meets our requirements. If I don't meet the requirements, I should get an error message.

password requirement:
- [ ] at least 8 characters

### LINKED ISSUE #700: UX for login page
Flow at https://invis.io/FHJ8CLVVSBK#/298613522_MIT_OPEN_IAFLOW_052118-pdf_1

Designs are at https://invis.io/JCJ5RJK9RA2#/297321219_Log_In_Step_1
([project link](https://projects.invisionapp.com/d/main#/console/14225859/297321219))

### acceptance criteria
- [ ] user can login with email and password
- [ ] if user enters an address ending with `mit.edu` display message and ask them to login with Touchstone
- [ ] if user enters an address for an existing account (e.g. edX) display error and ask them to login with any connected social auth
- [ ] if email / password combo is unrecognized, display and error message

### out of scope (related issues)
- "reveal password" eyeball button 
- #701 sign up / "Make an account with the email above" 
- #721 forgot password
- #671 user can login with Touchstone
- #713 user can login with edX

### LINKED ISSUE #701: UX for sign up page
Flow at https://invis.io/FHJ8CLVVSBK#/298613522_MIT_OPEN_IAFLOW_052118-pdf_1

Designs are at https://projects.invisionapp.com/d/main#/console/14225859/297321219

### acceptance criteria
- [ ] sign up with email
- [ ] error message if account already exists with that e-mail: offer any existing auth methods
- [ ] minimal e-mail validation (@ sign and at least one dot? or can we do something real-time with mailgun?)
- [ ] send verification mail 
- [ ] when user verifies email request password, request full name, password and T&C agreement
- [ ] if verification has expired, display an error and allow user to send another verification email
- [ ] #676 minimal password validation: at least 8 characters?
- [ ] error message if T&C are not agreed to
- [ ] return user to page they were viewing, with a toast-like message

### out of scope (related issues)
- #700 UX for login page
- forgot password
- #671 user can sign up with Touchstone
- user can sign up with edX
- "reveal password" eyeball button

### PR DESCRIPTION
**Pull request by rhysyngsun**
#### What are the relevant tickets?
Fixes #700 
Fixes #701 
Fixes #676 

#### What's this PR do?
This adds UI for login and registration via email

#### How should this be manually tested?

- Go to /register and confirm you can register (receive confirmation email, fill in details, etc)
- Go to /login and verify you can login
- Go to /register and try to register with the user you created, confirm you are redirected to enter a password instead
- Go to /login and enter a nonexistent email, confirm you are prompted to create an account and clicking the button sends the email and allows you to complete that flow.

#### Where should the reviewer start?
(Optional)

#### Any background context you want to provide?
(Optional)

#### Screenshots (if appropriate)
(Optional)

#### What GIF best describes this PR or how it makes you feel?
(Optional)

### CODE REVIEW

**Comment by noisecapella** [comment]
@rhysyngsun I can start looking over this now but do you know what is left to be done before the PR is fully ready for review?

**Comment by rhysyngsun** [comment]
@noisecapella just testing of the `static/js/containers/auth/*` pages, there might be some refactoring there to make testing easier, but it shouldn't affect other code.

**Comment by noisecapella** [comment]
I did successfully register and login but I'm seeing these issues when testing edge cases:

 - I managed to trigger a missing profile error when registering an account. I'll try to reproduce it tomorrow so I can provide steps to reproduce.
 - On this one email confirmation link I'm waiting indefinitely for the dialog to redirect me to fill in my profile: 
screenshot from 2018-06-28 18-36-27
 - A user can register two times with the same email address. They can register, click the link to activate the profile, then go back and register again, then fill in the profile again. If you then log out you are unable to log in with that account again.
 - Is there any error handling UI? If I log in with the incorrect password it doesn't tell me anything.
 - Regarding error handling, I'm seeing that I get a 200 with an error if I login with an incorrect password via `/login`. But if I login with `/register`, enter an existing email and then an incorrect password I get a 400 error. I don't think that's a blocker but it would be good to know why there's a difference there

Other minor issues or questions which can be dealt with separately:
 - Email confirmation toast disappears too quickly
 - If I have a 401 error from the REST API because the token expired it directs me to `/auth_required/`, even if I'm at `/login/` or `/register/`
 - I noticed we show the 'Sign out' link in the dropdown even if the user is already logged out
 - Are we holding off on login and register links until later?

**Comment by rhysyngsun** [comment]
On the minor issues, all of those except the 401 issue should get handled in separate PRs. I'll take a look at the other main issues you mentioned. If you're able to reproduce the unresponsive  page again could you give me a snapshot of the redux state at that point so I can see what happened?

**Comment by rhysyngsun** [comment]
@noisecapella I've addressed your primary feedback, except for 2 issues:

- **Couldn't reproduce**: On this one email confirmation link I'm waiting indefinitely for the dialog to redirect me to fill in my profile
- **I'd like to deal with this in a separate backend PR as this PR is already heavy enough and is feature gated**: A user can register two times with the same email address. They can register, click the link to activate the profile, then go back and register again, then fill in the profile again. If you then log out you are unable to log in with that account again.

**Comment by noisecapella** [comment]
I'm fine dealing with those issues separately. Can you create issues for them, or should I do it?

**Comment by rhysyngsun** [comment]
@noisecapella I'll create an issue for the second one, could you create one for the first with the steps to reproduce since I haven't been able to figure it out?

**Comment by noisecapella** [comment]
When I registered a new email address the response says "There is already an account with this email" even though it doesn't exist in my database

**Comment by noisecapella** [comment]
Nevermind, I think I'm looking at the wrong tables.

**Comment by noisecapella** [comment]
Functionality looks great now! 

This is not a blocker but there's strange validation UI behavior you can reproduce this way:
 - Login with your email
 - At the password prompt type 'xyz'. You should see a validation message that the password needs to be at least 8 characters
 - Type in 'xyzxyzxyz' (assuming not your actual password). It shows the validation message for the password but leaves the 8 character validation message behind

screenshot from 2018-07-03 16-54-57

**Comment by rhysyngsun** [comment]
Hmm, that probably needs some extra padding too

File: authentication/serializers.py [hunk 1]
```diff
@@ -44,13 +47,12 @@ def _authenticate(self, is_login=None):
...
             'request': request,
+            'flow': flow,
         }
 
-        if is_login is not None:
-            kwargs['is_login'] = is_login
-
         partial = partial_pipeline_data(backend, user, **kwargs)
         if partial:
+            print(partial.kwargs)
```
Comment (noisecapella) [inline authentication/serializers.py]: Stray print

File: authentication/views_test.py [hunk 2]
```diff
@@ -33,84 +44,322 @@ def test_auth_views_disabled(settings, client, user, url):
...
-    settings.FEATURES[features.EMAIL_AUTH] = True
-    user.set_password(user_password)
-    user.is_active = is_active
-    user.save()
-
-    if is_authenticated:
-        client.force_login(user)
+def last_partial_token():
+    """Returns the most recent partial token"""
+    return Partial.objects.last().token
```
Comment (noisecapella) [inline authentication/views_test.py]: Do we need an `order_by`? The `Partial` table doesn't seem to have a default ordering

**Comment by noisecapella** [comment]
One other non-blocking validation UI thing, the user gets a validation error right after clicking on the confirmation link. We should wait until the form is submitted, or at least an input field is updated, before having a validation error

screenshot from 2018-07-03 17-28-13

**Comment by noisecapella** [comment]
If I click on the terms and services link I'm directed to the terms of services on micromasters. When I click the back button and fill in the form it fails silently with a 400 error, "this field may not be null". I guess it lost track of the token from the URL when I went to the other page

File: static/js/components/auth/AuthDetailsForm.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        type="password"
+        name="password"
+        value={form.password || ""}
+        onChange={onUpdate}
+      />
+      {validationMessage(validation.password)}
+    </div>
+    <div className="tos row">
+      By creating an account, I agree to the{" "}
+      <a href={SETTINGS.authenticated_site.tos_url}>terms & conditions</a>{" "}
```
Comment (noisecapella) [inline static/js/components/auth/AuthDetailsForm.js]: Can we add an `target="_blank"` here with `rel="noopener noreferrer"`?

Comment (noisecapella) [reply static/js/components/auth/AuthDetailsForm.js]: Not a python comment but relates to my previous comment

**Comment by rhysyngsun** [comment]
@noisecapella ok, I've addressed those last two issues, I'll rebase as soon at those look good because it'll be easier to do so after I squash.

**Comment by noisecapella** [comment]
The validation looks good but the other comments aren't addressed yet. Feel free to rebase, I've looked at all the commits already

File: authentication/views_test.py (hunk 2 above)
Comment (rhysyngsun) [reply authentication/views_test.py]: this is unused now, so i'll just remove the whole func

**Comment by rhysyngsun** [comment]
@noisecapella feedback addressed, you can see it all in the second commit.

**Comment by noisecapella** [comment]
If I get a 500 error when registering an email (for example if the mail API fails) I don't see any error UI in the frontend. I don't remember if we decided to handle that in a separate PR or this one

File: static/js/containers/auth/LoginPasswordPage_test.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      }
+    )
+  })
+
+  afterEach(() => {
+    helper.cleanup()
+  })
+
+  //
+  ;[
```
[the remark this answers is not in the document]
Comment (rhysyngsun) [reply static/js/containers/auth/LoginPasswordPage_test.js]: I'm not certain if it's necessary, but it's consistent with our other usages so I'd prefer the consistency for now and figure out if we need it or not in another issue.

**Comment by rhysyngsun** [comment]
Let's deal with the 500 error separately, I'll put up an issue for it.

**Comment by rhysyngsun** [comment]
@noisecapella all set

**Comment by noisecapella** [comment]
Can you comment on what's happening with this change? https://github.com/mitodl/open-discussions/pull/824#discussion_r200405253

**Comment by noisecapella** [comment]
The link at https://github.com/mitodl/open-discussions/pull/824#discussion_r200405253 says we can use the `raf` package as a shim. What's the advantage to mocking it out ourselves?

**Comment by rhysyngsun** [comment]
@noisecapella actually, it appears the new tests run fine without this for whatever reason...so I just removed it. It looks like `enzyme` requires it, so I was probably just running into some weird case with that lib that isn't present anymore.

**Pull request merged**